tools/lsp/common.rs

```diff
@@ -25,32 +28,142 @@ pub type UrlVersion = Option<i32>;
 #[cfg(target_arch = "wasm32")]
 use crate::wasm_prelude::*;
 
-pub fn uri_to_file(uri: &lsp_types::Url) -> Option<PathBuf> {
-    let path = uri.to_file_path().ok()?;
-    let cleaned_path = i_slint_compiler::pathutils::clean_path(&path);
-    Some(cleaned_path)
+pub fn uri_to_file(uri: &Url) -> Option<PathBuf> {
+    if uri.scheme() == "builtin" {
+        let path = String::from("builtin:") + uri.path();
+        Some(PathBuf::from(path))
+    } else {
+        let path = uri.to_file_path().ok()?;
+        let cleaned_path = i_slint_compiler::pathutils::clean_path(&path);
+        Some(cleaned_path)
+    }
 }
 
-/// A cache of loaded documents
-pub struct DocumentCache {
-    pub(crate) documents: TypeLoader,
+pub fn file_to_uri(path: &Path) -> Option<Url> {
+    if path.starts_with("builtin:/") {
+        let p_str = path.to_string_lossy();
+        let p_str = if &p_str[9..11] == "///" {
```

Comment:
what's that?  That seems dangerous! what if we ever have non-ascii in our buildtin filename and 9..11 isn't right. not sure what you're trying to do there but that feels wrong.
Maybe you want to use strip_prefix ?